File: SDE_sheet/Arrays/08_merge_intervals/solution.go
func merge(intervals [][]int) [][]int {
	sort.Slice(intervals, func(i, j int) bool {
		return intervals[i][0] < intervals[j][0]
	})
	
	res := [][]int{}
	
	for i := 0; i < len(intervals); i++ {
		start, end := intervals[i][0], intervals[i][1]
		
		for i + 1 < len(intervals) && end >= intervals[i + 1][0] {
			i++
			end = max(end, intervals[i][1])
		}
		
		res = append(res, []int{start, end})
	}
	
	return res
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

File: SDE_sheet/Arrays/08_merge_intervals/solution.rs
impl Solution {
    fn merge(intervals: Vec<Vec<i32>>) -> Vec<Vec<i32>> {
        let mut intervals = intervals;
        intervals.sort_by(|a, b| {
            if a[0] != b[0] {
                a[0].cmp(&b[0])
            } else {
                a[1].cmp(&b[1])
            }
        });

        let mut result = vec![intervals[0].clone()];

        for interval in intervals.iter() {
            let last_end = result.last().unwrap()[1];
            if interval[0] <= last_end {
                let last = result.last_mut().unwrap();
                last[1] = std::cmp::max(last_end, interval[1]);
            } else {
                result.push(interval.clone());
            }
        }

        result
    }
}
